core/src/main/scala/sttp/tapir/metrics/Metric.scala

```diff
@@ -0,0 +1,15 @@
+package sttp.tapir.metrics
+
+import sttp.tapir.model.{ServerRequest, ServerResponse}
+
+case class Metric[F[_], M](
+    metric: M,
+    onRequest: Option[(ServerRequest, M) => F[Unit]] = None,
+    onResponse: Option[(ServerRequest, ServerResponse[_], M) => F[Unit]] = None,
+    onDecodeFailure: Option[(ServerRequest, M) => F[Unit]] = None
+) {
+  def onRequest(f: (ServerRequest, M) => F[Unit]): Metric[F, M] = copy(onRequest = Some(f))
```

Comment:
I think we should pass in here an implicit MonadError here, as in EndpointInterceptor - that way users don't have to provide one, when creating e.g. PrometheusMetrics. This is supposed to be mainly an internal thing (the MonadError)